sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -441,6 +537,17 @@ class Dataset[T] private(
   /** Collects the elements to an Array. */
   def collect(): Array[T] = rdd.collect()
 
+  /**
+   * (Java-specific)
+   * Collects the elements to a Java list.
+   *
+   * Due to the incompatibility problem between Scala and Java, the return type of [[collect()]] at
```

Comment:
Will the class tag do the trick? I tried to define a generic class with ClassTag:
```
class MyTest[T : ClassTag] {
  def t(): Array[T] = null
}

object MyTest {
  def apply[T](cls: Class[T]): MyTest[T] = {
    new MyTest[T]()(ClassTag(cls))
  }
}
```

The return type of `MyClass.t()` is still `Object` at java side.
I also tried to use scala RDD at java side, the return type of `RDD.collect()` is also `Object`.

One possible solution is to define `T <: AnyRef`, but I think it's hard to make it for `Dataset` or `RDD`.